core/src/main/scala/org/broadinstitute/dsde/rawls/deltalayer/DeltaLayerTranslator.scala

```diff
@@ -0,0 +1,132 @@
+    }
+
+    if (others.nonEmpty)
+      // in the future, may want to provide more info such as entity type/name and illegal operation
+      throw new DeltaLayerException(ERR_INVALID_OPERATION, code = BadRequest)
+
+    if (upserts.isEmpty)
+      // how did we reach here? This should be impossible
+      throw new DeltaLayerException(ERR_EMPTY_OPERATIONS, code = BadRequest)
+
+    // validate that all upsert values are string/boolean/number only
+    val allValues = allOps.collect {
+      case a:AddUpdateAttribute => a.addUpdateAttribute
+    }
+
+    if (allValues.size != allOps.size)
+      throw new DeltaLayerException("Unexpected error: count of update values did not match count of update operations",
+        code = InternalServerError)
+
+    val (supportedTypes, unsupportedTypes) = allValues partition isSupportedDataType
+
+    if (unsupportedTypes.nonEmpty)
+      throw new DeltaLayerException(ERR_INVALID_DATATYPE, code = BadRequest)
+
+    if (supportedTypes.isEmpty)
+      // how did we reach here? This should be impossible
+      throw new DeltaLayerException(ERR_INVALID_DATATYPE, code = BadRequest)
+
+    // everything validated; return the original update definitions
+    entityUpdates
+  }
+
+  /**
+   * is the supplied Attribute supported by Delta Layer?
+   * @param attr the Attribute to inspect
+   * @return whether or not Delta Layer supports this Attribute's type
+   */
+  private def isSupportedDataType(attr: Attribute): Boolean = attr match {
+    case AttributeBoolean(_) => true
+    case AttributeNumber(_) => true
+    case AttributeString(_) => true
+    case AttributeNull => true
+    case AttributeValueEmptyList => true
+    case AttributeValueList(elems) => elems forall isSupportedDataType
+    case _ =>
+      // AttributeNull, AttributeValueRawJson, AttributeEntityReference,
+      // AttributeEntityReferenceList, AttributeEntityReferenceEmptyList
+      false
+  }
+
+
+  /**
+   * Transforms the Rawls user-visible model classes into the models we will write
+   * into Delta Layer insert files. Calls validateEntityUpdates before performing
+   * transformations
+   *
+   * @param entityUpdates the updates to translate
+   * @return the collection of DeltaRow objects to write to a Delta Layer insert file
+   */
+  def translateEntityUpdates(entityUpdates: Seq[EntityUpdateDefinition]): Seq[DeltaRow] = {
+
+    val validUpdates = validateEntityUpdates(entityUpdates)
+
+    validUpdates.flatMap { update =>
+      // we expect all calls to use datarepo_row_id as entity name
+      val datarepoRowId = Try(UUID.fromString(update.name)) match {
+        case Success(id) => id
+        case Failure(_) => throw new DeltaLayerException(ERR_INVALID_ENTITYNAME, code = BadRequest)
+      }
+      update.operations map { op =>
+        // strip "default." from attr name
+        val name = op.name.namespace match {
+          case AttributeName.defaultNamespace => op.name.name
+          case _ => toDelimitedName(op.name)
```

Comment:
I _think_ you can just do 
```
val name = toDelimitedName(op.name)
```
since `toDelimitedName(...)` contains the same logic for the `defaultNamespace` case.

Reply:
switched! I *_swear_* I investigated `toDelimitedName` and thought it didn't do what I needed, but … it absolutely does. I also added a unit test for this, since it didn't have coverage.